File: e2e-test/flink-agents-end-to-end-tests-integration/src/test/java/org/apache/flink/agents/integration/test/ChatModelIntegrationAgent.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.integration.test;

import org.apache.flink.agents.api.Agent;
import org.apache.flink.agents.api.InputEvent;
import org.apache.flink.agents.api.OutputEvent;
import org.apache.flink.agents.api.annotation.Action;
import org.apache.flink.agents.api.annotation.ChatModelConnection;
import org.apache.flink.agents.api.annotation.ChatModelSetup;
import org.apache.flink.agents.api.annotation.Tool;
import org.apache.flink.agents.api.annotation.ToolParam;
import org.apache.flink.agents.api.chat.messages.ChatMessage;
import org.apache.flink.agents.api.chat.messages.MessageRole;
import org.apache.flink.agents.api.context.RunnerContext;
import org.apache.flink.agents.api.event.ChatRequestEvent;
import org.apache.flink.agents.api.event.ChatResponseEvent;
import org.apache.flink.agents.api.resource.ResourceDescriptor;
import org.apache.flink.agents.api.resource.ResourceType;
import org.apache.flink.agents.integrations.chatmodels.azureai.AzureAIChatModelConnection;
import org.apache.flink.agents.integrations.chatmodels.azureai.AzureAIChatModelSetup;
import org.apache.flink.agents.integrations.chatmodels.ollama.OllamaChatModelConnection;
import org.apache.flink.agents.integrations.chatmodels.ollama.OllamaChatModelSetup;
import org.apache.flink.agents.integrations.chatmodels.openai.OpenAIChatModelConnection;
import org.apache.flink.agents.integrations.chatmodels.openai.OpenAIChatModelSetup;

import java.util.Collections;
import java.util.List;

/**
 * Agent example that integrates an external Ollama chat model into Flink Agents.
 *
 * <p>This class demonstrates how to:
 *
 * <ul>
 *   <li>Declare a chat model connection using {@link ChatModelConnection} metadata pointing to
 *       {@link OllamaChatModelConnection}
 *   <li>Declare a chat model setup using {@link ChatModelSetup} metadata pointing to {@link
 *       OllamaChatModelSetup}
 *   <li>Expose callable tools via {@link Tool} annotated static methods (temperature conversion,
 *       BMI, random number)
 *   <li>Fetch a chat model from the {@link RunnerContext} and perform a single-turn chat
 *   <li>Emit the model response as an {@link OutputEvent}
 * </ul>
 *
 * <p>The {@code ollamaChatModel()} method publishes a resource with type {@link
 * ResourceType#CHAT_MODEL} so it can be retrieved at runtime inside the {@code process} action. The
 * resource is configured with the connection name, the model name and the list of tool names that
 * the model is allowed to call.
 */
public class ChatModelIntegrationAgent extends Agent {
    public static final String OLLAMA_MODEL = "qwen3:0.6b";

    @ChatModelConnection
    public static ResourceDescriptor chatModelConnection() {
        String provider = System.getProperty("MODEL_PROVIDER", "OLLAMA");
        if (provider.equals("OLLAMA")) {
            return ResourceDescriptor.Builder.newBuilder(OllamaChatModelConnection.class.getName())
                    .addInitialArgument("endpoint", "http://localhost:11434")
                    .addInitialArgument("requestTimeout", 240)
                    .build();
        } else if (provider.equals("AZURE")) {
            String endpoint = System.getenv().get("AZURE_ENDPOINT");
            String apiKey = System.getenv().get("AZURE_API_KEY");
            return ResourceDescriptor.Builder.newBuilder(AzureAIChatModelConnection.class.getName())
                    .addInitialArgument("endpoint", endpoint)
                    .addInitialArgument("apiKey", apiKey)
                    .build();
        } else if (provider.equals("OPENAI")) {
            String apiKey = System.getenv().get("OPENAI_API_KEY");
            return ResourceDescriptor.Builder.newBuilder(OpenAIChatModelConnection.class.getName())
                    .addInitialArgument("api_key", apiKey)
                    .build();
        } else {
            throw new RuntimeException(String.format("Unknown model provider %s", provider));
        }
    }

    @ChatModelSetup
    public static ResourceDescriptor chatModel() {
        String provider = System.getProperty("MODEL_PROVIDER", "OLLAMA");

        if (provider.equals("OLLAMA")) {
            return ResourceDescriptor.Builder.newBuilder(OllamaChatModelSetup.class.getName())
                    .addInitialArgument("connection", "chatModelConnection")
                    .addInitialArgument("model", OLLAMA_MODEL)
                    .addInitialArgument(
                            "tools",
                            List.of("calculateBMI", "convertTemperature", "createRandomNumber"))
                    .build();
        } else if (provider.equals("AZURE")) {
            return ResourceDescriptor.Builder.newBuilder(AzureAIChatModelSetup.class.getName())
                    .addInitialArgument("connection", "chatModelConnection")
                    .addInitialArgument("model", "gpt-4o")
                    .addInitialArgument(
                            "tools",
                            List.of("calculateBMI", "convertTemperature", "createRandomNumber"))
                    .build();
        } else if (provider.equals("OPENAI")) {
            return ResourceDescriptor.Builder.newBuilder(OpenAIChatModelSetup.class.getName())
                    .addInitialArgument("connection", "chatModelConnection")
                    .addInitialArgument("model", "gpt-4o-mini")
                    .addInitialArgument(
                            "tools",
                            List.of("calculateBMI", "convertTemperature", "createRandomNumber"))
                    .build();
        } else {
            throw new RuntimeException(String.format("Unknown model provider %s", provider));
        }
    }

    @Tool(description = "Converts temperature between Celsius and Fahrenheit")
    public static double convertTemperature(
            @ToolParam(name = "value", description = "Temperature value to convert") Double value,
            @ToolParam(
                            name = "fromUnit",
                            description = "Source unit ('C' for Celsius or 'F' for Fahrenheit)")
                    String fromUnit,
            @ToolParam(
                            name = "toUnit",
                            description = "Target unit ('C' for Celsius or 'F' for Fahrenheit)")
                    String toUnit) {

        fromUnit = fromUnit.toUpperCase();
        toUnit = toUnit.toUpperCase();

        if (fromUnit.equals(toUnit)) {
            return value;
        }

        if (fromUnit.equals("C") && toUnit.equals("F")) {
            return (value * 9 / 5) + 32;
        } else if (fromUnit.equals("F") && toUnit.equals("C")) {
            return (value - 32) * 5 / 9;
        } else {
            throw new IllegalArgumentException("Invalid temperature units. Use 'C' or 'F'");
        }
    }

    @Tool(description = "Calculates Body Mass Index (BMI)")
    public static double calculateBMI(
            @ToolParam(name = "weightKg", description = "Weight in kilograms") Double weightKg,
            @ToolParam(name = "heightM", description = "Height in meters") Double heightM) {

        if (weightKg <= 0 || heightM <= 0) {
            throw new IllegalArgumentException("Weight and height must be positive values");
        }
        return weightKg / (heightM * heightM);
    }

    @Tool(description = "Create a random number")
    public static double createRandomNumber() {
        return Math.random();
    }

    @Action(listenEvents = {InputEvent.class})
    public static void process(InputEvent event, RunnerContext ctx) throws Exception {
        ctx.sendEvent(
                new ChatRequestEvent(
                        "chatModel",
                        Collections.singletonList(
                                new ChatMessage(MessageRole.USER, (String) event.getInput()))));
    }

    @Action(listenEvents = {ChatResponseEvent.class})
    public static void processChatResponse(ChatResponseEvent event, RunnerContext ctx) {
        ctx.sendEvent(new OutputEvent(event.getResponse().getContent()));
    }
}

File: python/flink_agents/e2e_tests/e2e_tests_integration/chat_model_integration_agent.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
import os

from flink_agents.api.agents.agent import Agent
from flink_agents.api.chat_message import ChatMessage, MessageRole
from flink_agents.api.decorators import (
    action,
    chat_model_connection,
    chat_model_setup,
    tool,
)
from flink_agents.api.events.chat_event import ChatRequestEvent, ChatResponseEvent
from flink_agents.api.events.event import InputEvent, OutputEvent
from flink_agents.api.resource import (
    ResourceDescriptor,
    ResourceName,
)
from flink_agents.api.runner_context import RunnerContext


class ChatModelTestAgent(Agent):
    """Example agent demonstrating the new ChatModel architecture."""

    @chat_model_connection
    @staticmethod
    def openai_connection() -> ResourceDescriptor:
        """ChatModelConnection responsible for openai model service connection."""
        return ResourceDescriptor(
            clazz=ResourceName.ChatModel.OPENAI_CONNECTION, api_key=os.environ.get("OPENAI_API_KEY")
        )

    @chat_model_connection
    @staticmethod
    def azure_openai_connection() -> ResourceDescriptor:
        """ChatModelConnection responsible for openai model service connection."""
        return ResourceDescriptor(
            clazz=ResourceName.ChatModel.AZURE_OPENAI_CONNECTION,
            api_key=os.environ.get("AZURE_OPENAI_API_KEY"),
            api_version=os.environ.get("AZURE_OPENAI_API_VERSION"),
            azure_endpoint=os.environ.get("AZURE_OPENAI_ENDPOINT"),
        )

    @chat_model_connection
    @staticmethod
    def tongyi_connection() -> ResourceDescriptor:
        """ChatModelConnection responsible for tongyi model service connection."""
        return ResourceDescriptor(clazz=ResourceName.ChatModel.TONGYI_CONNECTION)

    @chat_model_connection
    @staticmethod
    def ollama_connection() -> ResourceDescriptor:
        """ChatModelConnection responsible for ollama model service connection."""
        return ResourceDescriptor(
            clazz=ResourceName.ChatModel.OLLAMA_CONNECTION, request_timeout=240.0
        )

    @chat_model_setup
    @staticmethod
    def math_chat_model() -> ResourceDescriptor:
        """ChatModel which focus on math, and reuse ChatModelConnection."""
        model_provider = os.environ.get("MODEL_PROVIDER")
        if model_provider == "Tongyi":
            return ResourceDescriptor(
                clazz=ResourceName.ChatModel.TONGYI_SETUP,
                connection="tongyi_connection",
                model=os.environ.get("TONGYI_CHAT_MODEL", "qwen-plus"),
                tools=["add"],
            )
        elif model_provider == "Ollama":
            return ResourceDescriptor(
                clazz=ResourceName.ChatModel.OLLAMA_SETUP,
                connection="ollama_connection",
                model=os.environ.get("OLLAMA_CHAT_MODEL", "qwen3:1.7b"),
                tools=["add"],
                extract_reasoning=True,
            )
        elif model_provider == "OpenAI":
            return ResourceDescriptor(
                clazz=ResourceName.ChatModel.OPENAI_SETUP,
                connection="openai_connection",
                model=os.environ.get("OPENAI_CHAT_MODEL", "gpt-3.5-turbo"),
                tools=["add"],
            )
        elif model_provider == "AzureOpenAI":
            return ResourceDescriptor(
                clazz=ResourceName.ChatModel.AZURE_OPENAI_SETUP,
                connection="azure_openai_connection",
                model=os.environ.get("AZURE_OPENAI_CHAT_MODEL", "gpt-5"),
                tools=["add"],
            )
        else:
            err_msg = f"Unknown model_provider {model_provider}"
            raise RuntimeError(err_msg)

    @chat_model_setup
    @staticmethod
    def creative_chat_model() -> ResourceDescriptor:
        """ChatModel which focus on text generate, and reuse ChatModelConnection."""
        model_provider = os.environ.get("MODEL_PROVIDER")
        if model_provider == "Tongyi":
            return ResourceDescriptor(
                clazz=ResourceName.ChatModel.TONGYI_SETUP,
                connection="tongyi_connection",
                model=os.environ.get("TONGYI_CHAT_MODEL", "qwen-plus"),
            )
        elif model_provider == "Ollama":
            return ResourceDescriptor(
                clazz=ResourceName.ChatModel.OLLAMA_SETUP,
                connection="ollama_connection",
                model=os.environ.get("OLLAMA_CHAT_MODEL", "qwen3:1.7b"),
                extract_reasoning=True,
            )
        elif model_provider == "OpenAI":
            return ResourceDescriptor(
                clazz=ResourceName.ChatModel.OPENAI_SETUP,
                connection="openai_connection",
                model=os.environ.get("OPENAI_CHAT_MODEL", "gpt-3.5-turbo"),
            )
        elif model_provider == "AzureOpenAI":
            return ResourceDescriptor(
                clazz=ResourceName.ChatModel.AZURE_OPENAI_SETUP,
                connection="azure_openai_connection",
                model=os.environ.get("AZURE_OPENAI_CHAT_MODEL", "gpt-5"),
            )
        else:
            err_msg = f"Unknown model_provider {model_provider}"
            raise RuntimeError(err_msg)

    @tool
    @staticmethod
    def add(a: int, b: int) -> int:
        """Calculate the sum of a and b.

        Parameters
        ----------
        a : int
            The first operand
        b : int
            The second operand

        Returns:
        -------
        int:
            The sum of a and b
        """
        return a + b

    @action(InputEvent)
    @staticmethod
    def process_input(event: InputEvent, ctx: RunnerContext) -> None:
        """User defined action for processing input.

        In this action, we will send ChatRequestEvent to trigger built-in actions.
        """
        input_text = event.input.lower()
        model_name = (
            "math_chat_model"
            if ("calculate" in input_text or "sum" in input_text)
            else "creative_chat_model"
        )
        ctx.send_event(
            ChatRequestEvent(
                model=model_name,
                messages=[ChatMessage(role=MessageRole.USER, content=event.input)],
            )
        )

    @action(ChatResponseEvent)
    @staticmethod
    def process_chat_response(event: ChatResponseEvent, ctx: RunnerContext) -> None:
        """User defined action for processing chat model response."""
        input = event.response
        if event.response and input.content:
            ctx.send_event(OutputEvent(output=input.content))
